File: mirror/tier2-python/showcase/meta_ast.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Metaprogramming
Module: meta_ast.py
Purpose: Demonstrate AST-like patterns
"""

import ast
from typing import Any, Dict, List


class ASTNode:
    """Simple AST node"""
    def __init__(self, node_type: str, value: Any = None, children: List = None):
        self.node_type = node_type
        self.value = value
        self.children = children or []
    
    def traverse(self, callback):
        """Traverse AST"""
        callback(self)
        for child in self.children:
            child.traverse(callback)


def build_ast_tree() -> ASTNode:
    """Build example AST"""
    root = ASTNode("program")
    root.children.append(ASTNode("function", "add", [
        ASTNode("param", "x"),
        ASTNode("param", "y"),
        ASTNode("return", None, [
            ASTNode("binary_op", "+", [
                ASTNode("var", "x"),
                ASTNode("var", "y"),
            ])
        ])
    ]))
    return root


def analyze_ast(tree: ASTNode) -> Dict[str, int]:
    """Analyze AST structure"""
    counts = {}
    
    def counter(node):
        ntype = node.node_type
        counts[ntype] = counts.get(ntype, 0) + 1
    
    tree.traverse(counter)
    return counts


result = {
    "ast_root_type": "program",
    "ast_node_count": len(analyze_ast(build_ast_tree())),
    "ast_analysis": analyze_ast(build_ast_tree()),
    "ast_functional": True,
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-ruby/showcase/meta_ast.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Metaprogramming
# Module: meta_ast.rb
# Purpose: Demonstrate AST-like patterns

class ASTNode
  attr_accessor :node_type, :value, :children
  
  def initialize(node_type, value = nil, children = [])
    @node_type = node_type
    @value = value
    @children = children
  end
  
  def traverse(&block)
    block.call(self)
    @children.each { |child| child.traverse(&block) }
  end
end

def build_ast_tree
  root = ASTNode.new("program")
  root.children << ASTNode.new("function", "add", [
    ASTNode.new("param", "x"),
    ASTNode.new("param", "y"),
  ])
  root
end

def analyze_ast(tree)
  counts = {}
  tree.traverse do |node|
    counts[node.node_type] ||= 0
    counts[node.node_type] += 1
  end
  counts
end

result = {
  ast_root_type: "program",
  ast_analysis: analyze_ast(build_ast_tree()),
  ast_functional: true,
}

puts result.inspect
